odl/space/fspace.py

```diff
@@ -154,196 +356,346 @@ class FunctionSetVector(Operator):
+        vectorized : bool, optional
+            Whether the function supports vectorized evaluation
         """
         if not isinstance(fset, FunctionSet):
             raise TypeError('function set {!r} not a `FunctionSet` '
                             'instance.'.format(fset))
-
-        if fcall is None and fapply is None:
-            raise ValueError('call function and apply function cannot '
-                             'both be `None`.')
-
-        if fcall is not None and not callable(fcall):
-            raise TypeError('call function {!r} is not callable.'
-                            ''.format(fcall))
-
-        if fapply is not None and not callable(fapply):
-            raise TypeError('apply function {!r} is not callable.'
-                            ''.format(fapply))
+        if not callable(fcall):
+            raise TypeError('function {!r} is not callable.'.format(fcall))
 
         self._space = fset
-        self._call = fcall
-        self._apply = fapply
+        super().__init__(self._space.domain, self._space.range, linear=False)
+
+        # Determine which type of implementation fcall is
+        if isinstance(fcall, FunctionSetVector):
+            call_has_out, call_out_optional, _ = _dispatch_call_args(
+                bound_call=fcall._call)
+        elif isinstance(fcall, np.ufunc):
+            if fcall.nin != 1:
+                raise ValueError('cannot use `ufunc` with more than one '
+                                 'input parameter.')
+            call_has_out = call_out_optional = True
+        elif isfunction(fcall):
+            call_has_out, call_out_optional, _ = _dispatch_call_args(
+                unbound_call=fcall)
+        elif isclass(fcall):
+            call_has_out, call_out_optional, _ = _dispatch_call_args(
+                bound_call=fcall.__call__)
+        else:
+            raise TypeError('callable type {!r} not understood.')
+
+        self._call_has_out = call_has_out
+        self._call_out_optional = call_out_optional
+
+        if not call_has_out:
+            # Out-of-place only
+            def ip_wrapper(func):
+                @wraps(func)
+                def wrapper(x, out, **kwargs):
+                    return func(self, x, out, **kwargs)
+                return wrapper
+
+            self._call_in_place = ip_wrapper(_default_in_place)
+            self._call_out_of_place = fcall
+        elif call_out_optional:
+            # Dual-use
+            self._call_in_place = self._call_out_of_place = fcall
+        else:
+            # In-place only
+            self._call_in_place = fcall
+            # No way to safely determine a data type for the output array,
+            # therefore not out-of-place default applicable
+            self._call_out_of_place = _out_of_place_not_impl
 
-        # Todo: allow users to specify linear
-        super().__init__(self.space.domain, self.space.range, linear=False)
+        self._vectorized = bool(vectorized)
```

Comment:
It's the discouraged use case anyway (we should clearly discourage it in the doc then), so it doesn't hurt so much if we make it even slower :-)